include/config/config_interface.h

```diff
@@ -0,0 +1,113 @@
+// We use a concept instead of virtual methods, because the context of
+// its usage (getConfig) is using a template anyways, due to the
+// different return types (RuntimeConfig vs.
+// PreprocessingConfig). Alternatively, making the constructor private
+// and instead creating a factory method would also be possible.
+template <typename C>
+concept Config = requires(C c, const C cc, const VerifiedConfigAttributes& config_source) {
+   { C::getConfigSpecification() } -> std::same_as<ConfigSpecification>;
+
+   /// Vector of config files that the user gave (or that is provided
+   /// by the type via its defaults) that should be loaded and shadowed
+   /// in the order of the vector.
+   { cc.getConfigPaths() } -> std::same_as<std::vector<std::filesystem::path>>;
+
+   /// Overwrite the fields of an instance of the target type; done
+   /// that way so that multiple kinds of config sources can shadow
+   /// each other's values by application in sequence. Does not throw
+   /// exceptions, except overwriteFrom can call SILO_PANIC when there
+   /// is an inconsistency (bug) between ConfigSpecification and
+   /// overwriteFrom implementation.
+   { c.overwriteFrom(config_source) } -> std::same_as<void>;
+
+   /// Validation / Sanity checks about the values of this config
+   { c.validate() } -> std::same_as<void>;
+};
+
+/// This function needs a reference to the (remaining) command line
+/// arguments to be parsed, thus the application gets a chance to take
+/// off some arguments first (like "api" or "preprocessing" in the
+/// current SILO version).
+///
+/// In case of error, returns the exit code that the caller should
+/// pass to exit(): 0 if the user gave --help, 1 in case of erroneous
+/// usage (the error is already printed in that case).
+template <Config C>
+std::variant<C, int32_t> getConfig(
+   std::span<const std::string> cmd,
+   const std::vector<std::string>& allow_list_for_env_vars
+) {
+   const auto config_specification = C::getConfigSpecification();
+   try {
+      auto cmd_source = CommandLineArguments{cmd}.verify(config_specification);
+      if (cmd_source.asks_for_help) {
+         std::cout << config_specification.helpText() << "\n" << std::flush;
+         return 0;
+      }
+      if (!cmd_source.positional_arguments.empty()) {
+         throw silo::config::ConfigException{"SILO does not expect positional arguments"};
+      }
+
+      auto env_source =
+         EnvironmentVariables::newWithAllowListAndEnv(allow_list_for_env_vars, environ)
+            .verify(config_specification);
+
+      // First, get the config paths, if any
+      SPDLOG_TRACE("Now overwriting config from defaults");
+      C config;
+      SPDLOG_TRACE("Now overwriting config from environment variables");
+      config.overwriteFrom(env_source);
+      SPDLOG_TRACE("Now overwriting config from command line arguments");
+      config.overwriteFrom(cmd_source);
+      auto config_paths = config.getConfigPaths();
```

Comment:
I think I would restructure this logically. It's weird that the configs have their config values and the config file paths in a single class. IMO it would make more sense to separate them. I imagine that this part of the code could looks like:
```suggestion
      C config;
      auto config_paths = config.getConfigFilePaths(cmd_source, env_source);
```

Then, internally the config classes could have some mechanism:
```c++
   std::vector<std::filesystem::path> result;
   default_runtime_config = getConfigPath(defaultRuntimeConfigOptionKey(), cmd_source, env_source);
   if (default_runtime_config.has_value()) {
      result.emplace_back(default_runtime_config.value());
   }
   runtime_config = getConfigPath(runtimeConfigOptionKey(), cmd_source, env_source);
   if (runtime_config.has_value()) {
      result.emplace_back(runtime_config.value());
   }
   return result;
``` 

That would streamline the logic:
* the configs still specify the files that they expect to read
* the configs don't carry file paths as members anymore (which is weird - IMO the config class should mirror the config file structure)
* It reflects the logic that you need cmd_source and env_source to get config file paths.
* This method here becomes cleaner - we don't need to parse config twice